File: examples/external-binary-search-C/BinarySearchTest.c

#include <stdlib.h>
#include <stdio.h>
#include "common.h"
#include "Record.h"
#include "ExternalBinarySearch.h"

char *program;
const int MAX_KEY = 100000;

double getMilliseconds();

void conduct_random_searches(FILE *dataFile, long int n)
{
	long int i;
	unsigned long int key;
	Record *record;
	char *buffer;

	for (i=0; i<n; i++) {
		key = (unsigned long int) random() % MAX_KEY;
		if (DEBUG >= 1) {
				printf("\nSearching for key = %ld\n\n", key);
		}
		record = extBinarySearch(dataFile, key);
		if (DEBUG >= 1) {
			if (record) {
				buffer = toString(record);
				printf("%s\n", buffer);
				printf("found: record with key %ld\n", key);
				free(record);
				free(buffer);
			} else {
				printf("not found: record with key %ld\n", key);
			}
		}
	}
}

int main (int argc, char **argv)
{
	int n;
	long int count;
	FILE *dataFile;
	double startTime, totalTime;

	program = argv[0];

	if (argc != 3) {
		fprintf(stderr, "Usage: %s <data file name> <number of searches>\n",
				program);
		exit(1);
	}
	n = atoi(argv[2]);
	dataFile = fopen(argv[1], "r");
	if (!dataFile) {
		perror(program);
		exit(1);
	}
	fseek(dataFile, 0, SEEK_END);
	count = ftell(dataFile);
	count = count/sizeof(Record);
	if (DEBUG >= 1) {
		   fprintf(stderr, "data file has %ld records\n", count);
	}

	startTime = getMilliseconds();
	conduct_random_searches(dataFile, n);
	totalTime = getMilliseconds() - startTime;
	printf("Elapsed time for %d searches = %8.2f seconds\n", n, totalTime/1000.0);

	fclose(dataFile);
	exit(0);
}

/* vim: set ts=4: */

File: examples/external-binary-search/BinarySearchTest.java
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.channels.FileChannel;
import java.util.Random;

public class BinarySearchTest
{
	private int DEBUG = 0;

	private void conductRandomSearches(FileChannel dataFile, int n, int MAX_KEY)
	throws IOException
	{
		int i;
		int key;
	    Record record;
		Random generator = new Random();
		ExternalBinarySearch searcher = new ExternalBinarySearch();
		searcher.setDebug(DEBUG);

		for (i=0; i<n; i++) {
			key = generator.nextInt(MAX_KEY);
			if (DEBUG >= 1) {
				System.out.println("\nSearching for key = " + key + " \n");
			}
			record = searcher.search(dataFile, key);
			if (DEBUG >= 1) {
				if (record != null) {
					System.out.println(record);
					System.out.println("found: record with key "+key);
				} else {
					System.out.println("not found!: record with key "+key);
				}
			}
		}
	}


	public static void main (String [] argv)
	{
		long count;
		BinarySearchTest tester = new BinarySearchTest();
		Record record = new Record();

		if (argv.length != 3) {
			System.err.println("Usage: java BinarySearchTest <data file name> <number of searches> <max key>");
			System.exit(1);
		}
		int n = Integer.parseInt(argv[1]);
		int maxKey = Integer.parseInt(argv[2]);
		try {
			FileChannel dataFile = new RandomAccessFile(argv[0],"r").getChannel();

			count = dataFile.size()/(long ) record.getDiskSize();
			if (tester.DEBUG >= 1) {
		   		System.err.println("data file has " + count + "  records");
			}

			long startTime = System.currentTimeMillis();
			tester.conductRandomSearches(dataFile, n, maxKey);
			long totalTime = System.currentTimeMillis() - startTime;

			System.out.println("Elapsed time for " + n + " searches = " + totalTime/1000.0 
								+ " seconds"); 
			dataFile.close();
			System.exit(0);
		} catch (IOException e) {
		   	System.err.println(e);
			System.exit(1);
		}
	}
}
